Replace ERR_EXPLAIN in scene/* code

KoBeWi opened the pull request:
Addresses part of #31207

Aside from straightforward adding `_MSG `, I changed some `if` conditions to `FAIL_COND_*` macros. There are however two `EXPLAIN`s I didn't know what to do about, as they come with some font-specific macro (which doesn't have `MSG` version):
https://github.com/godotengine/godot/blob/d2a67c9c1fdee470064f2b3c5750c98f174b5399/scene/resources/dynamic_font.cpp#L187-L188
(and below)

akien-mga commented on scene/2d/area_2d.cpp:
```diff
@@ -401,10 +398,7 @@ void Area2D::set_monitoring(bool p_enable) {
-		ERR_EXPLAIN("Function blocked during in/out signal. Use set_deferred(\"monitoring\",true/false)");
-	}
-	ERR_FAIL_COND(locked);
+	ERR_FAIL_COND_MSG(locked, "Function blocked during in/out signal. Use set_deferred(\"monitoring\",true/false)");
```
Nitpick, but while we're changing this, can you add a space after `"monitoring",`? Same below.

akien-mga commented on scene/2d/physics_body_2d.cpp:
```diff
@@ -905,10 +891,7 @@ void RigidBody2D::set_contact_monitor(bool p_enabled) {
-			ERR_EXPLAIN("Can't disable contact monitoring during in/out callback. Use call_deferred(\"set_contact_monitor\",false) instead");
-		}
-		ERR_FAIL_COND(contact_monitor->locked);
+		ERR_FAIL_COND_MSG(contact_monitor->locked, "Can't disable contact monitoring during in/out callback. Use call_deferred(\"set_contact_monitor\",false) instead");
```
Same here.

akien-mga commented:
Only checked a few of them so far, will check the rest tomorrow.

While we're editing those, might be worth make the style of error messages consistent. Some end with dots and some don't, I think it should likely always end with a dot?

KoBeWi committed:
Remove ERR_EXPLAIN from scene/* code

KoBeWi commented:
Added dots everywhere.

And what about translations? Shouldn't errors be translated? 4 out of the 100+ I changed were translated (some font-related stuff).

akien-mga commented:
> And what about translations? Shouldn't errors be translated? 4 out of the 100+ I changed were translated (some font-related stuff).

Errors shouldn't be translated, they often point to engine issues and we need them in English when reported on GitHub.

akien-mga commented on scene/resources/dynamic_font.cpp:
```diff
@@ -118,8 +118,7 @@ Error DynamicFontAtSize::_load() {
 
-	ERR_EXPLAIN(TTR("Error initializing FreeType."));
-	ERR_FAIL_COND_V(error != 0, ERR_CANT_CREATE);
+	ERR_FAIL_COND_V_MSG(error != 0, ERR_CANT_CREATE, TTR("Error initializing FreeType."));
```
The `TTR()` calls are bogus in this file, you can remove them.
`TTR` should never be used outside editor code, it gets replaced by `String()` on non-tools builds.

akien-mga commented on scene/animation/animation_cache.cpp:
```diff
@@ -80,8 +80,7 @@ void AnimationCache::_update_cache() {
 			path_cache.push_back(Path());
-			ERR_EXPLAIN("Invalid Track Path in Animation: " + np);
-			ERR_CONTINUE(!node);
+			ERR_CONTINUE_MSG(!node, "Invalid Track Path in Animation: " + np + ".");
```
I'd lowercase "track path in animation" for style consistency, but it's a nitpick, I don't mind if it stays as is :)

akien-mga commented on scene/animation/tween.cpp:
```diff
@@ -1185,44 +1185,38 @@ bool Tween::_build_interpolation(InterpolateType p_interpolation_type, Object *p
-	ERR_EXPLAIN("Initial value type does not match final value type!"); // TODO: Print both types to make debugging easier
-	ERR_FAIL_COND_V(p_initial_val.get_type() != p_final_val.get_type(), false); // Do the initial and final value types match?
+	// TODO: Print both types to make debugging easier
+	ERR_FAIL_COND_V_MSG(p_initial_val.get_type() != p_final_val.get_type(), false, "Initial value type does not match final value type."); // Do the initial and final value types match?
```
Let's solve the TODO:
```
"Initial value type '" + Variant::get_type_name(p_initial_val.get_type()) + "' does not match final value type '" + Variant::get_type_name(p_final_val.get_type()) + "'."
```
(Untested, but if it compiles it should be fine I guess.)

akien-mga replied:
The trailing comments in this class are a bit useless BTW, they just state exactly what the test + explanatory message say... but well, it's unrelated to this PR :P

KoBeWi replied:
I was going to solve it, but wasn't sure how to get the type string.
(also removed the comments xd)

akien-mga commented on scene/animation/tween.cpp:
```diff
@@ -1231,8 +1225,8 @@ bool Tween::_build_interpolation(InterpolateType p_interpolation_type, Object *p
 		// Does the object even have the requested method?
-		ERR_EXPLAIN("Tween target object has no method named: " + *p_method); // TODO: Fix this error message
-		ERR_FAIL_COND_V(!p_object->has_method(*p_method), false);
+		// TODO: Fix this error message
```
I don't see what's wrong with that error message, so I'd just drop the cryptic TODO.

akien-mga commented on scene/animation/tween.cpp:
```diff
@@ -1301,8 +1295,7 @@ bool Tween::interpolate_callback(Object *p_object, real_t p_duration, String p_c
 	// Check whether the object even has the callback
-	ERR_EXPLAIN("Object has no callback named: %s" + p_callback);
-	ERR_FAIL_COND_V(!p_object->has_method(p_callback), false);
+	ERR_FAIL_COND_V_MSG(!p_object->has_method(p_callback), false, "Object has no callback named: %s" + p_callback + ".");
```
These errors were (and still are) bogus. There's no substitution done. This was meant to be passed to `vformat` I guess, but here you can simply drop the `%s` (same below).

akien-mga commented on scene/gui/control.cpp:
```diff
@@ -2377,12 +2367,7 @@ Control *Control::_get_focus_neighbour(Margin p_margin, int p_count) {
-				ERR_EXPLAIN("Neighbour focus node is not a control: " + n->get_name());
-				ERR_FAIL_V(NULL);
-			}
+			ERR_FAIL_COND_V_MSG(!c, NULL, "Neighbour focus node is not a control: " + n->get_name() + ".");
```
en_GB spotted, should be `Neighbor`.

KoBeWi replied:
The method itself has "neighbour" in name...

akien-mga replied:
Oops :D Probably one to list in #16863.

akien-mga commented on scene/main/node.cpp:
```diff
@@ -1239,10 +1218,7 @@ void Node::_propagate_validate_owner() {
-		ERR_EXPLAIN("Parent node is busy setting up children, remove_node() failed. Consider using call_deferred(\"remove_child\",child) instead.");
-		ERR_FAIL_COND(data.blocked > 0);
-	}
+	ERR_FAIL_COND_MSG(data.blocked > 0, "Parent node is busy setting up children, remove_node() failed. Consider using call_deferred(\"remove_child\",child) instead.");
```
And space after `"remove_child",`.

akien-mga commented on scene/resources/text_file.cpp:
```diff
@@ -67,10 +67,7 @@ Error TextFile::load_text(const String &p_path) {
-		ERR_EXPLAIN("Script '" + p_path + "' contains invalid unicode (utf-8), so it was not loaded. Please ensure that scripts are saved in valid utf-8 unicode.");
-		ERR_FAIL_V(ERR_INVALID_DATA);
-	}
+	ERR_FAIL_COND_V_MSG(s.parse_utf8((const char *)w.ptr()), ERR_INVALID_DATA, "Script '" + p_path + "' contains invalid unicode (utf-8), so it was not loaded. Please ensure that scripts are saved in valid utf-8 unicode.");
```
While I'm nitpicking, maybe capitalize "UTF-8" here? (two occurrences)

akien-mga commented:
Checked everything, seems good. Left a few nitpicks, once solved it can be merged.

KoBeWi commented:
I resolved the nitpicks.

akien-mga commented:
Thanks!
